cpp/dolfinx/fem/assemble_expression_impl.h

```diff
@@ -0,0 +1,126 @@
+// Copyright (C) 2019-2020 Garth N. Wells
+//
+// This file is part of DOLFINx (https://www.fenicsproject.org)
+//
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+#pragma once
+
+#include "Expression.h"
+#include "FunctionSpace.h"
+#include "traits.h"
+#include "utils.h"
+#include <algorithm>
+#include <dolfinx/common/IndexMap.h>
+#include <dolfinx/mesh/Geometry.h>
+#include <dolfinx/mesh/Mesh.h>
+#include <dolfinx/mesh/Topology.h>
+#include <memory>
+#include <vector>
+
+namespace dolfinx::fem::impl
+{
+/// @brief
+/// @tparam T
+/// @tparam U
+/// @param values
+/// @param vshape
+/// @param fn
+/// @param Xshape
+/// @param value_size
+/// @param coeffs
+/// @param cstride
+/// @param constant_data
+/// @param mesh
+/// @param entities
+/// @param V
+template <dolfinx::scalar T, std::floating_point U>
+void assemble_expression(
+    std::span<T> values, std::array<std::size_t, 2> vshape,
+    fem::FEkernel<T> auto fn, std::array<std::size_t, 2> Xshape,
+    std::size_t value_size, std::span<const T> coeffs, std::size_t cstride,
+    std::span<const T> constant_data, const mesh::Mesh<U>& mesh,
+    std::span<const std::int32_t> entities,
+    std::optional<std::reference_wrapper<const FunctionSpace<U>>> V)
+{
+  namespace md = MDSPAN_IMPL_STANDARD_NAMESPACE;
+
+  std::size_t estride;
+  if (mesh.topology()->dim() == Xshape[1])
+    estride = 1;
+  else if (mesh.topology()->dim() == Xshape[1] + 1)
+    estride = 2;
+  else
+    throw std::runtime_error("Invalid dimension of evaluation points.");
```

Comment:
Since we are doing this rewrite. Should we think about how one could tabulate an expression on an interior facet?
The abstraction with checking the shape of `X` does not work, as an interior facet would have the same dimension for the reference points. 
Do we need a similar enum to `IntegralType`, or just use it?

Reply:
Let's leave this to another PR.